pilot/pkg/xds/eds.go

```diff
@@ -249,6 +240,35 @@ func (s *DiscoveryServer) deleteService(cluster, serviceName, namespace string)
 	}
 }
 
+// UpdateServiceAccount updates the service endpoints' sa when service/endpoint event happens.
+func (s *DiscoveryServer) UpdateServiceAccount(shards *EndpointShards, clusterID, serviceName string, endpoints []*model.IstioEndpoint) bool {
+	oldServiceAccount := shards.ServiceAccounts
+	shards.mutex.Lock()
+	defer shards.mutex.Unlock()
+
+	// Check if ServiceAccounts have changed. We should do a full push if they have changed.
+	serviceAccounts := sets.Set{}
+	for cluster, epShards := range shards.Shards {
+		if cluster == clusterID {
+			epShards = endpoints
```

Comment:
We are updating istioEndpoints for this cluster with `clusterID`, so here we are calculating all sas with these new endpoints rather than the old ones.